internal/controller/clientprofile_controller.go

```diff
@@ -0,0 +1,337 @@
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=clientprofile/finalizers,verbs=update
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=cephconnection,verbs=get;list;watch;update
+//+kubebuilder:rbac:groups="",resources=configmap,verbs=get;list;watch;create;update;patch;delete
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ClientProfileReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Filter update events based on metadata.generation changes, will filter events
+	// for non-spec changes on most resource types.
+	genChangedPredicate := predicate.GenerationChangedPredicate{}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1a1.ClientProfile{}).
+		Owns(
+			&csiv1a1.CephConnection{},
+			builder.MatchEveryOwner,
+			builder.WithPredicates(genChangedPredicate),
+		).
+		Owns(
+			&corev1.ConfigMap{},
+			builder.MatchEveryOwner,
+			builder.WithPredicates(genChangedPredicate),
+		).
+		Complete(r)
+}
+
+func (r *ClientProfileReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx)
+	log.Info("Starting reconcile iteration for ClientProfile", "req", req)
+
+	reconcileHandler := ClientProfileReconcile{}
+	reconcileHandler.ClientProfileReconciler = *r
+	reconcileHandler.ctx = ctx
+	reconcileHandler.log = log
+	reconcileHandler.clientProfile.Name = req.Name
+	reconcileHandler.clientProfile.Namespace = req.Namespace
+
+	err := reconcileHandler.reconcile()
+	if err != nil {
+		log.Error(err, "ClientProfile reconciliation failed")
+	} else {
+		log.Info("ClientProfile reconciliation completed successfully")
+	}
+	return ctrl.Result{}, err
+}
+
+func (r *ClientProfileReconcile) reconcile() error {
+	if err := r.loadAndValidate(); err != nil {
+		return err
+	}
+
+	// Ensure a finalizer on the ClientProfile to allow proper clean up
+	if ctrlutil.AddFinalizer(&r.clientProfile, cleanupFinalizer) {
+		if err := r.Update(r.ctx, &r.clientProfile); err != nil {
+			r.log.Error(err, "Failed to add a cleanup finalizer on ClientProfile")
+			return err
+		}
+	}
+
+	if err := r.reconcileCephConnection(); err != nil {
+		return err
+	}
+	if err := r.reconcileCephCsiClusterInfo(); err != nil {
+		return err
+	}
+
+	if r.cleanUp {
+		ctrlutil.RemoveFinalizer(&r.clientProfile, cleanupFinalizer)
+		if err := r.Update(r.ctx, &r.clientProfile); err != nil {
+			r.log.Error(err, "Failed to add a cleanup finalizer on config resource")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *ClientProfileReconcile) loadAndValidate() error {
+	// Load the ClientProfile
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&r.clientProfile), &r.clientProfile); err != nil {
+		r.log.Error(err, "Failed loading ClientProfile")
+		return err
+	}
+	r.cleanUp = r.clientProfile.DeletionTimestamp != nil
+
+	// Validate a pointer to a ceph cluster resource
+	if r.clientProfile.Spec.CephConnectionRef.Name == "" {
```

Comment:
any reason in not using a CEL rule?

Reply:
A CEL rule is not a replacement for double-checking in code. 
Are we missing a CEL rule?

Reply:
I think we are missing CEL. xxRef struct is made required but the field Name in LocalRef is optional for backwards compat and so CEL rule on our struct would be good.

Reply:
I am not aware the CEL can look into structs that we do not define (in this case it is `LocalObjectReference` object)
If there is a way to achieve that, please point me in the write direction